File: src/backtracking/lc_79_word_search/word_search.py
class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        visited = set()
        def dfs(i,j,k):
            if (i,j) in visited:
                return False
            visited.add((i,j))
            if k > len(word) or i < 0 or j < 0 or i >= len(board) or j >= len(board[0]):
                return False
            if k == len(word):
                return True
            #check row
            if i+1 < len(board) and word[k] == board[i+1][j]:
                if dfs(i+1, j, k+1):
                    return True 
            #check col
            if j+1 < len(board[0]) and word[k] == board[i][j+1]:
                if dfs(i, j+1, k+1):
                    return True 

            #check row back
            if i-1 >= 0 and word[k] == board[i-1][j]:
                if dfs(i-1, j, k+1):
                    return True 

            #check col back
            if j-1 >= 0 and word[k] == board[i][j-1]:
                if dfs(i, j-1, k+1):
                    return True 

            visited.remove((i,j))
            return False
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    if dfs(i,j,1):
                        return True
                    visited = set()
        return False

File: src/backtracking/lc_79_word_search/word_search.java
import java.util.*;
class Solution {

    public int lenrow;
    public int lencol;
    public Set<String> visited = new HashSet<>();
    public int length;

    public boolean exist(char[][] board, String word) {
        this.lenrow = board.length;
        this.lencol = board[0].length;
        this.length = word.length();
        
        for(int i = 0; i < lenrow; i++) {
            for(int j = 0; j < lencol; j++) {
                if (board[i][j] == word.charAt(0)) {
                    if(bt(i,j,1,board, word)) 
                        return true;
                    visited.clear();
                }
            }    
        }
        return false;
    }

    public boolean bt(int i, int j, int k, char[][] board, String word) {
        if (visited.contains(i + "," + j))
            return false;
        visited.add(i + "," + j);
        if (k > length || i < 0 || j < 0 || i >= lenrow || j >= lencol)
            return false;
        if (k == length)
            return true;
        if(i+1 < lenrow && board[i+1][j] == word.charAt(k)) {
            if(bt(i+1, j, k+1,board, word))
                return true;
        }
        if(j+1 < lencol && board[i][j+1] == word.charAt(k)) {
            if(bt(i, j+1, k+1,board, word))
                return true;
        }
        if(i-1 >= 0 && board[i-1][j] == word.charAt(k)) {
            if(bt(i-1, j, k+1,board, word))
                return true;
        }
        if(j-1 >= 0 && board[i][j-1] == word.charAt(k)) {
            if(bt(i, j-1, k+1,board, word))
                return true;
        }
        
        visited.remove(i + "," + j);
        return false;
    }
}